### PR TITLE
Support storing multiple mobile numbers and email addresses per user

### PR DESCRIPTION
**Pull request by PasinduYeshan**
### Proposed changes in this pull request

>$subject. 

This feature will be disabled by default for the time being, however with the following config in `deployment.toml`, the multiple support for email addresses and mobile numbers can be enabled.

```
[identity_mgt.user_claim_update]
enable_multiple_emails_and_mobile_numbers = true
```
verifiedEmailAddresses and verifiedMobileNumbers claims can only be updated when `Enable user email verification on update` and the above config are both enabled.

### Multiple email addresses support

Users will be able to store multiple email addresses  in newly introduced custom claim http://wso2.org/claims/emailAddresses and multiple verified email addresses in http://wso2.org/claims/verifiedEmailAddresses claim.

### Multiple mobile numbers support

Users will be able to store multiple mobile numbers in newly introduced custom claim http://wso2.org/claims/mobileNumbers and multiple verified mobile numbers in http://wso2.org/claims/verifiedMobileNumbers claim.

#### 1. **Handling Primary Email/Mobile Updates When Multi-attribute Support is Disabled**
   - **Previous Behavior**: When multi-attribute support was disabled, changing the primary email or mobile number skipped verification if the new number/email was already in the verified list.
   - **Resolution**: The verification logic has been refined. Now, if multi-attribute support is disabled, existing verified numbers or emails are not considered, ensuring proper verification steps. This logic has been moved inside the multi-attribute-specific flow.

#### 2. **Ensuring Proper Error Handling with Disabled Verification & Enabled Multi-attribute Support**
   - **Previous Behavior**: When verification was disabled but multi-attribute support was enabled, updating the primary mobile number with a new, unlisted number threw an error.
   - **Resolution**: The system now automatically adds the new mobile number to the mobile numbers list, preventing the error and allowing the update to proceed smoothly.

#### 3. **Automatic Addition of Primary Number to Verified Lists with Enabled Multi-attribute & Verification**
   - **Previous Behavior**: The existing primary mobile number was not being added to the `mobileNumbers` and `verifiedMobileNumbers` lists during updates.
   - **Resolution**: The old primary mobile number is now automatically added to both the `mobileNumbers` and `verifiedMobileNumbers` lists, ensuring the data remains consistent.

#### 4. **Improved Handling of New Primary Number with Enabled Multi-attribute & Verification**
   - **Previous Behavior**: When adding a new number as the primary mobile number, an error was thrown due to the number not being recognized in the verified mobile numbers lists.
   - **Resolution**: The new primary number is now added to the `mobileNumbers` and `verifiedMobileNumbers` lists, without throwing an error ensuring a smoother transition.

#### 5. **Streamlining Verification Logic for Pre-verified Mobile Numbers**
   - **Previous Behavior**: When updating a mobile number already in the verified list, the system did not follow the newly introduced `SKIP_ON_ALREADY_VERIFIED_EMAIL_ADDRESSES` logic but used the `SKIP_ON_INAPPLICABLE_CLAIMS` rule.
   - **Resolution**: The verification logic has been moved to the appropriate flow, ensuring that the system properly handles pre-verified numbers based on the correct logic.

#### 6. Add Unit test for following classes.
  - MobileNumberVerificationHandler
  - UserEmailVerificationHandler
  - UserSelfRegistrationManager
  - Utils

## Related Git Issue
- https://github.com/wso2/product-is/issues/19991

### Referenced PRs
- https://github.com/wso2-extensions/identity-governance/pull/817

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java [hunk 1]
```diff
@@ -84,9 +92,21 @@ public void handleEvent(Event event) throws IdentityEventException {
                         getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME));
         Map<String, String> claims = (Map<String, String>) eventProperties.get(IdentityEventConstants.EventProperty
                 .USER_CLAIMS);
+        if (claims == null) {
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java:95]: If claims map is null, do we have to go through the rest of the method? Can't we add a return statement here?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java [hunk 2]
```diff
@@ -88,16 +97,31 @@ public void handleEvent(Event event) throws IdentityEventException {
 
         Map<String, String> claims = (Map<String, String>) eventProperties.get(IdentityEventConstants.EventProperty
                 .USER_CLAIMS);
+        if (claims == null) {
+            claims = new HashMap<>();
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java:101]: Same as in MobileNumberVerificationHandler can't we add a return statement here?

Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java:101]: No, we can't add a return statement here, as the same handler also processes POST_SET_USER_CLAIMS events where no claims are required. The reason I added this null check is to ensure null safety, as in some cases, claim values are accessed directly without proper null checks.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java (hunk 1 above)
Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java:95]: No, we can't add a return statement here, as the same handler also processes POST_SET_USER_CLAIMS events where no claims are required. The reason I added this null check is to ensure null safety, as in some cases, claim values are accessed directly without proper null checks.

**Review by lashinijay — DISMISSED** [review]

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 3]
```diff
@@ -759,12 +760,39 @@ private UserRecoveryData validateSelfRegistrationCode(String code, String verifi
...
+                    try {
+                        List<String> verifiedEmails = Utils.getMultiValuedClaim(userStoreManager, user,
+                                IdentityRecoveryConstants.VERIFIED_EMAIL_ADDRESSES_CLAIM);
+                        verifiedEmails.add(pendingEmailClaimValue);
+                        userClaims.put(IdentityRecoveryConstants.VERIFIED_EMAIL_ADDRESSES_CLAIM, StringUtils.join(
+                                verifiedEmails, multiAttributeSeparator));
+
+                        List<String> allEmails = Utils.getMultiValuedClaim(userStoreManager, user,
+                                IdentityRecoveryConstants.EMAIL_ADDRESSES_CLAIM);
+                        if (!allEmails.contains(pendingEmailClaimValue)) {
```
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:781]: Can allEmails be null?

**Review by sadilchamishka — APPROVED** [review]

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java (hunk 3 above)
Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:781]: No `Utils.getMultiValuedClaim` always return ArrayList.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/util/Utils.java [hunk 4]
```diff
@@ -127,6 +130,18 @@ public class Utils {
      */
     private static ThreadLocal<String> skipSendingSmsOtpVerificationOnUpdate = new ThreadLocal<>();
 
+    /**
+     * This thread local variable is used to stop verification pending mobile number from being set as the primary
+     * mobile number when only updating the verifiedMobileNumbers claim.
+     */
+    private static ThreadLocal<Boolean> isOnlyVerifiedMobileNumbersUpdated = new ThreadLocal<>();
```
[the remark this answers is not in the document]
Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/util/Utils.java:137]: As discussed, the ThreadLocal unsetting logic was added to the PRE_ADD_USER_CLAIM events before the value change.

**Review by sadilchamishka — APPROVED** [review]

**Review by lashinijay — APPROVED** [review]

**Pull request merged**